### PR TITLE
chore: create tslint rule to allow @HostListener and @HostBinding in abstract classes

### PR DESCRIPTION
**Pull request by josephperrott**

### CODE REVIEW

File: tools/tslint-rules/noHostDecoratorInConcreteRule.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import * as ts from 'typescript';
+import * as Lint from 'tslint';
+
+export class Rule extends Lint.Rules.AbstractRule {
+    apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
```
Comment (crisbeto) [inline tools/tslint-rules/noHostDecoratorInConcreteRule.ts]: It looks like the indentation of the whole file is too large (seems like it's a 4, but it should be 2).

File: tools/tslint-rules/noHostDecoratorInConcreteRule.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    visitClassDeclaration(node: ts.ClassDeclaration) {
+      if (!node.modifiers || !this.getOptions().length) { return; }
+
+      // Do not check the class if its abstract.
+      if (!!node.modifiers.find(modifier => modifier.kind === ts.SyntaxKind.AbstractKeyword)) {
+        return;
+      }
+
+      node.members.forEach(el => {
+        if (el.decorators) {
```
Comment (crisbeto) [inline tools/tslint-rules/noHostDecoratorInConcreteRule.ts]: This block could be a bit more concise with a `filter` and a `map`:

```ts
node.members
  .filter(el => !!el.decorators)
  .map(el => el.decorators)
  .forEach(decorator => {
    const decoratorText: string = decorator.getChildAt(1).getText();
    const matchedDecorator: string = this.getOptions().find(
      (item: string) => decoratorText.startsWith(item));
    
      if (!!matchedDecorator) {
        this.addFailureFromStartToEnd(decorator.getChildAt(1).pos - 1, decorator.end,
            `The @${matchedDecorator} decorator may only be used in abstract classes. In
              concrete classes use \`host\` in the component definition instead.`);
      }
});
```

File: tools/tslint-rules/noHostDecoratorInConcreteRule.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+          el.decorators.forEach(decorator => {
+            const decoratorText: string = decorator.getChildAt(1).getText();
+            const matchedDecorator: string = this.getOptions().find(
+              (item: string) => decoratorText.startsWith(item));
+            if (!!matchedDecorator) {
+              this.addFailureFromStartToEnd(decorator.getChildAt(1).pos - 1, decorator.end,
```
Comment (crisbeto) [inline tools/tslint-rules/noHostDecoratorInConcreteRule.ts]: Not completely sure, but you should be able to use `this.addFailureAtNode(decorator, ...)` instead.

Comment (josephperrott) [reply tools/tslint-rules/noHostDecoratorInConcreteRule.ts]: I initially tried this, but the starting position for the decorator was the character after the previous node, which made the tslint annotation go crazy.

File: tools/tslint-rules/noHostDecoratorInConcreteRule.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import * as ts from 'typescript';
+import * as Lint from 'tslint';
+
+export class Rule extends Lint.Rules.AbstractRule {
+  apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+    return this.applyWithWalker(new Walker(sourceFile, this.getOptions()));
+  }
+}
+
+class Walker extends Lint.RuleWalker {
```
Comment (crisbeto) [inline tools/tslint-rules/noHostDecoratorInConcreteRule.ts]: Add a short docstring about what the rule does?

**Pull request merged**